### ISSUE TITLE
Database design

### ISSUE THREAD

**Issue by matt-dale** [matt-dale/designdb#1]
**The Equipment Library**
The high level concept is to have a global Equipment Library that users may incorporate into their projects, but also allow them to create their own Equipment for use in only their projects.

**Comment by gmichael225** [comment]
Hey Matt,
Happy to try and come up with something for this.
Q: What're you using for hosting at the moment?

Michael

**Comment by matt-dale** [comment]
I'm using Heroku for hosting and Hover for domain management.

**Comment by matt-dale** [comment]
**The Global>User>Project level conundrum**
If you look at the database model for ProjectCable in the old system, you will notice that there is a hard ForeignKey Link to the Cable model.  This is how I handled project level vs. global level cables in the "cable library".  There are definitely some issues with this plan.    
First, if there are any changes to the global Cable instance that the Project Cable was referenced to(like deleting or changing the name), they will automatically be changed in the Project Level instance. The user may not like that.    
Second, the user may want a customized version of the global Cable instance and edit specific attributes of it.    
    
I propose a system where the Global and Project level items are decoupled.  The Equipment Library will function as templates for Project level Equipment, then the users can edit the Project Equipment in any way that they would like.  This also allows for Project level instances to be created that have no relation to the Global library for items that are custom, or too unique that they don't need to be in the Global library.    

There is one more layer to this that I'd like to design:  The User Level
An example:  The shop that I typically get my equipment from has JBL speakers.  The Global Library has this speaker with an NL4 connector on it, but the shop where I get my equipment provides it with a 1/4" connector.  Since I use this speaker a lot, I wouldn't want to create a Project level speaker for each project, the system should just "know" that when I add the JBL speaker from the Global library then it should have a 1/4" connector on it instead of an NL4.    
This example is a pipe dream and might be more hassle than it's worth...

**Comment by gmichael225** [comment]
Hey, haven't had a close look yet, but IMO the best way would be (pretty much what you said):
- Global cables/devices/etc act like "templates" in that when you add one to a design / project, the SQL row is effectively duplicated from the templates table into the project table. You can still maintain an FK to the global in case a user wants to "reset to defaults" or something.
Re: JBL's with 1/4" connector example, I'd imagine "JBL XXX with 1/4" connector" could be an item in a user library of sorts? So rather than whenever you add any JBL speaker it comes with 1/4", you have a set of JBL speakers in your "user library" that you've changed to have 1/4" connectors.


So, to sum up:
 - There are two types of templates (think: classes) - user templates and global templates.
 - Project cables / devices / etc are unique instances of these templates, in which all the template info is duplicated.

Thoughts?

**Comment by matt-dale** [comment]
Agreed. Makes sense.

**Comment by matt-dale** [comment]
**My Label Mistake**
So much of the project thus far has been about printing out physical labels.  What I failed to do in the Beta was create a Label table in the database that would hold the details about each label that needed to be printed.  I created a Label table globally for each type of label, but I just hardcoded all sorts of details about how the label would print into the label printing views.  I now realize that each cable end and piece of equipment and connection needs to have  LabelDetails instance created for it so that the user may override specifics of what information is printed on the label and what the layout should look like.

**Decision about the treatment of Cable**
From a "best practices" standpoint, I've been struggling with whether to treat Cable as a child of Equipment. If you think about equipment as an extremely general term, then Cable should fall into that but there are a lot of attributes about Cables that will dictate how designdb should operate.  I'm leaning more towards keeping it separate as it is in the Beta.  
But then again, if each Equipment instance has a dedicated Attributes instance, then a field in the Attributes table could be "EquipmentType" with a value of "Cable" or "Multi" or "Breakout", "Speaker", "Amplifier" etc.    
In production apps that I've used, I noticed a similar approach to this.  There is a general Attributes table with a GenericForeignKey that may relate to an instance from any other table.  It makes for a table with many fields encompassing a lot of different situations, but it seems that there is a reason for this instead of just adding the Attributes fields to the parent's table (maybe performance?).   

As I built it in the Beta, Cable and Equipment both have anywhere from 0 to many Connections and PatchPoints are generated by connecting one Connection to another.  (This code is not reflected in the "public" version on bitbucket.)

This seems to stay fairly clean and easy to follow.  To remedy the Label Mistake, each Connection will get a ProjectLabel instance created for it and each Equipment instance will get a ProjectLabel instance created for it.


Sorry for the stream of consciousness rant...

**Comment by matt-dale** [comment]
**Project Level Permissions System**
This should be fairly simple to implement with Django's Groups and a permission_required/user_passes_test decorator, but I overlooked this in the Beta.   
The idea is to allow a project owner to add collaborators to the project for read/write or read-only access. 

The collaborators will also be able to access the "show completion portal" which will be a digital form of the "show crew lists" from the Beta.  The target device for the portal will be a tablet to view completed signal flow docs, rack drawings and to check items off the "show crew lists".
